### PR TITLE
feat(auth): System e2e tests for access control and delegations

### PR DESCRIPTION
**Pull request by eirikurn**
## What

Create E2E tests to test access control and delegations.

## Why

More trust in releases.

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

### CODE REVIEW

**Review by oddsson — APPROVED** [review]

**Review by pshomov — APPROVED** [review]
Great work! 

Not to put the cart in front of the horse, but do we need to test our other locale?

**Review by pshomov — CHANGES_REQUESTED** [review]
On a second read - the last two tests for the delegation test are dependent on each other and I think should be merged into one. Seems the guidance is that this kind of tests is encouraged to be large with lots of assertions.

**Review by snaerseljan — APPROVED** [review]
Awesome work 🥇 . It would be nice that me our one of us would do a thorough  acceptance test when we are not under deadline.

File: apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
+import { BrowserContext, expect, test } from '@playwright/test'
+import { urls } from '../../../support/utils'
+import { session } from '../../../support/session'
+import { helpers } from '../../../support/locator-helpers'
+
+test.use({ baseURL: urls.islandisBaseUrl })
```
Comment (kristofer-lp8f) [inline apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts:6]: Possibly redundant?

File: apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      '[data-dialog-ref="access-delete-modal"] >> role=button[name*=Eyða]',
+    )
+    const count = await delegations.count()
+
+    for (let i = 0; i < count; ++i) {
+      await deleteButton.click()
+      await confirmButton.click()
+      await confirmButton.waitFor({ state: 'hidden' })
+    }
+
```
Comment (kristofer-lp8f) [inline apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts:50]: Check out https://playwright.dev/docs/api/class-locator#locator-evaluate-all

File: apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    await page.locator('data-testid=proceed').click()
+
+    const access = page.locator('input[type=checkbox][id*=scope]')
+    await expect(access).not.toHaveCount(0)
+    const accessCount = await access.count()
+    for (let i = 0; i < accessCount; i++) {
+      await access.nth(i).setChecked(true)
+    }
+    await page.locator('data-testid=proceed').click()
+    await page.locator('role=button[name=Staðfesta]').click()
```
Comment (kristofer-lp8f) [inline apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts:75]: Check out https://playwright.dev/docs/api/class-locator#locator-evaluate-all

**Review by kristofer-lp8f — COMMENTED** [review]
Note that the the E2E test you wrote is _not_ a smoke test, since it is _modifying_ state, i.e. _changing_ user delegations. The test should be placed in `service-portal/acceptance/`

I would suggest skipping `data-testid` and opting for selecting solely by roles.

File: apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      page.locator(
+        '[data-testid="access-card"], [data-testid="delegations-empty-state"]',
+      ),
+    ).not.toHaveCount(0)
+
+    // Act
+    const delegations = page
+      .locator('data-testid=access-card')
+      .filter({ hasText: 'Gervimaður Ameríku' })
+    const deleteButton = delegations.first().locator('role=button[name=Eyða]')
```
[the remark this answers is not in the document]
Comment (pshomov) [reply apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts:39]: `data-testid` is not recommended as a general way to target elements. In fact it comes last on the list. For more info - https://www.notion.so/General-information-c8191cb240f743128df40c0040fda5a0#b63b8d822ae34eeaa193f105914fbaae

File: apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts (hunk 2 above)
Comment (eirikurn) [reply apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts:50]: Yeah, I took this pattern from here:

https://playwright.dev/docs/locators#rare-use-cases

The locator API is much nicer in this use-case than writing the same logic with DOM nodes in a page function.

File: apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts (hunk 3 above)
Comment (eirikurn) [reply apps/system-e2e/src/tests/service-portal/smoke/access-control.spec.ts:75]: I admit `Locator#evaluateAll` might be a bit cleaner here, but I like staying in `Locator` land in Playwright tests.

File: apps/system-e2e/src/tests/service-portal/smoke/service-portal.spec.ts [hunk 5]
```diff
@@ -7,35 +7,61 @@ test.use({ baseURL: urls.islandisBaseUrl })
...
+    const { findByRole } = helpers(page)
+    await page.goto('/minarsidur')
+
+    // Act
+    await page.locator('data-testid=user-menu >> visible=true').click()
+    await page.locator('role=button[name="Skipta um notanda"]').click()
+    const firstCompany = page.locator('role=button[name*="Prókúra"]').first()
+    await expect(firstCompany).toBeVisible()
+    const companyName = await firstCompany
+      .locator('.identity-card--name')
```
[the remark this answers is not in the document]
Comment (eirikurn) [reply apps/system-e2e/src/tests/service-portal/smoke/service-portal.spec.ts:59]: Don't have anything better to select by for now, and this page is in the IDS repo.

**Review by kristofer-lp8f — APPROVED** [review]
LGTM

File: apps/system-e2e/src/tests/service-portal/acceptance/access-control.spec.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+  })
+
+  test.afterAll(async () => {
+    await context.close()
+  })
+
+  test('can remove, create and use custom delegations', async ({ browser }) => {
+    const page = await context.newPage()
+
+    await test.step('Remove delegations', async () => {
```
Comment (pshomov) [inline apps/system-e2e/src/tests/service-portal/acceptance/access-control.spec.ts:29]: Did not know about `test.step`. Very nice!

**Review by pshomov — APPROVED** [review]
LGTM

**Pull request merged**